src/clj/api/mutations.clj

```diff
@@ -359,3 +359,104 @@
+
+(defn- delete-bayes-factor-analysis! [{:keys [id db s3 bucket-name user-id]}]
+  (let [{:keys [log-file-url locations-file-url output-file-url]}
+        (bayes-factor-model/get-bayes-factor-analysis db {:id id})]
+    (doseq [url (remove nil? [log-file-url locations-file-url output-file-url])]
+      (delete-s3-object! {:url url :user-id user-id :s3 s3 :bucket bucket-name}))
+    (analysis-model/delete-analysis db {:id id})))
+
+(defn- delete-time-slicer-analysis! [{:keys [id db s3 bucket-name user-id]}]
+  (let [{:keys [trees-file-url output-file-url]} (time-slicer-model/get-time-slicer db {:id id})]
+    (log/info "delete-time-slicer-analysis" {:trees-file-url  trees-file-url
+                                             :output-file-url output-file-url})
+    (doseq [url (remove nil? [output-file-url trees-file-url])]
+      (delete-s3-object! {:url url :user-id user-id :s3 s3 :bucket bucket-name}))
+    (analysis-model/delete-analysis db {:id id})))
+
+;; TODO : add tests for this mutation
+(defn delete-analysis
+  [{:keys [authed-user-id db s3 bucket-name]} {id :id :as args} _]
+  (try
+    (let [_                 (log/info "delete-analysis" {:user/id authed-user-id
+                                                         :args    args})
+          {:keys [of-type]} (analysis-model/get-analysis db {:id id})
+          args-map          {:id id :db db :s3 s3 :bucket-name bucket-name :user-id authed-user-id}]
+      (case (keyword of-type)
+        :CONTINUOUS_TREE       (delete-continuous-tree-analysis! args-map)
+        :DISCRETE_TREE         (delete-discrete-tree-analysis! args-map)
+        :BAYES_FACTOR_ANALYSIS (delete-bayes-factor-analysis! args-map)
+        :TIME_SLICER           (delete-time-slicer-analysis! args-map)
+        (log/error "Unknown analysis type" {:of-type of-type :id id}))
+      {:id id})
+    (catch Exception e
+      (log/error "Exception occured when deleting analysis" {:analysis/id id
+                                                             :error       e}))))
+
+(defn delete-file [{:keys [authed-user-id s3 bucket-name]} {url :url :as args} _]
+  (try
+    (let [_          (log/info "delete-file" {:user/id authed-user-id
+                                              :args    args})
+          object-key (delete-s3-object! {:url url :user-id authed-user-id :s3 s3 :bucket bucket-name})]
+      {:key object-key})
+    (catch Exception e
+      (log/error "Exception occured when deleting file" {:url   url
+                                                         :error e}))))
+
+(defn delete-user-data!
+  [{:keys [user-id db s3 bucket-name]}]
+  (let [user-objects (:Contents (aws-s3/list-objects s3 {:bucket bucket-name
+                                                         :prefix user-id}))]
+    (aws-s3/delete-objects s3 {:bucket bucket-name :objects user-objects})
+    (analysis-model/delete-all-user-analysis db {:user-id user-id})))
+
+(defn delete-user-data
```

Comment:
I wanted to avoid repeating the same code since there is:

- a delete-data mutation that just cleans all the users data
- a delete-account mutation that deletes the data AND the account

Maybe I should make `delete-user-data!` private to mark that it's an implementation detail?

Reply:
Hmm I see, so one is the mutation wrapper and the other the implementation. Maybe making the mutations ns only contain wrappers and then putting other funcionts in other namespaces help, but whatever is ok for me.